core/profiles/src/main/java/software/amazon/awssdk/profiles/ProfileFileSupplier.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.profiles;
+
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.profiles.internal.ProfileFileRefresher;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+
+/**
+ * Encapsulates the logic for supplying either a single or multiple ProfileFile instances.
+ * <p>
+ * Each call to the {@link #get()} method will result in either a new or previously supplied profile based on the
+ * implementation's rules.
+ */
+@SdkPublicApi
+@FunctionalInterface
+public interface ProfileFileSupplier extends Supplier<ProfileFile>, SdkAutoCloseable {
+
+    @Override
+    default void close() {
+    }
+
+    /**
+     * Creates a {@link ProfileFileSupplier} capable of producing multiple profile objects by aggregating the default
+     * credentials and configuration files as determined by {@link ProfileFileLocation#credentialsFileLocation()} abd
+     * {@link ProfileFileLocation#configurationFileLocation()}. This supplier will return a new ProfileFile instance only once
+     * either disk file has been modified. Multiple calls to the supplier while both disk files are unchanged will return the
+     * same object.
+     *
+     * @return Implementation of {@link ProfileFileSupplier} that is capable of supplying a new aggregate profile when either file
+     *         has been modified.
+     */
+    static ProfileFileSupplier defaultSupplier() {
+        Optional<ProfileFileSupplier> credentialsSupplierOptional
+            = ProfileFileLocation.credentialsFileLocation()
+                                 .map(path -> reloadWhenModified(path, ProfileFile.Type.CREDENTIALS));
+
+        Optional<ProfileFileSupplier> configurationSupplierOptional
+            = ProfileFileLocation.configurationFileLocation()
+                                 .map(path -> reloadWhenModified(path, ProfileFile.Type.CONFIGURATION));
+
+        ProfileFileSupplier supplier = () -> null;
```

Comment:
Should we return an empty file instead of null? Or do we handle null everywhere that we handle profile files?